[TEP-0076]Pipeline results support array

Yongxuanzhang opened the pull request:
<!-- 🎉🎉🎉 Thank you for the PR!!! 🎉🎉🎉 -->

# Changes
This is part of work in [TEP-0076.](https://github.com/tektoncd/community/blob/main/teps/0076-array-result-types.md)
This commit update the PipelineResult and PipelineRunResult to uspport
array results, meanwhile support the whole array and array indexing
reference from PipelineResult to TaskResult. Before this commit the
pipeline level result only support string.

/kind feature

<!-- 
Describe your changes here- ideally you can get that description straight from your descriptive commit message(s)! 

In addition, categorize the changes you're making using the "/kind" Prow command, example:

/kind <kind>

Supported kinds are: bug, cleanup, design, documentation, feature, flake, misc, question, tep
-->

# Submitter Checklist

As the author of this PR, please check off the items in this checklist:

- [x] [Docs](https://github.com/tektoncd/community/blob/main/standards.md#docs) included if any changes are user facing
- [x] [Tests](https://github.com/tektoncd/community/blob/main/standards.md#tests) included if any functionality added or changed
- [x] Follows the [commit message standard](https://github.com/tektoncd/community/blob/main/standards.md#commits)
- [x] Meets the [Tekton contributor standards](https://github.com/tektoncd/community/blob/main/standards.md) (including
  functionality, content, code)
- [x] Release notes block below has been filled in
(if there are no user facing changes, use release note "NONE")

# Release Notes
```release-note
Support array results and indexing array for pipeline level as an alpha feature.

A pipeline can specify a type to create array result, such as:

    results:
      - name: echo-indexing-array-results
        type: string
        description: array element
        value: $(tasks.task1.results.array-results[1])
      - name: echo-array-results
        type: array
        description: whole array
        value: $(tasks.task2.results.array-results[*])

And the task script can populate result in an array form with:

echo -n "[\"hello\",\"world\"]" | tee $(results.array-results.path)

The pipeline results can refer to task results to collect them.
This feature is part of the TEP-0076. 
```
<!--
Describe any user facing changes here, or delete this block.

Examples of user facing changes:
- API changes
- Bug fixes
- Any changes in behavior
- Changes requiring upgrade notices or deprecation warnings

For pull requests with a release note:

```release-note
Your release note here
```

For pull requests that require additional action from users switching to the new release, include the string "action required" (case insensitive) in the release note:

```release-note
action required: your release note here
```

For pull requests that don't need to be mentioned at release time, use the `/release-note-none` Prow command to add the `release-note-none` label to the PR. You can also write the string "NONE" as a release note in your PR description:

```release-note
NONE
```
-->

Yongxuanzhang commented:
/test pull-tekton-pipeline-go-coverage

Yongxuanzhang commented:
/retest

Yongxuanzhang commented:
/retest

ywluogg commented:
/assign @ywluogg

lbernick commented on pkg/apis/pipeline/v1beta1/pipeline_types.go:
```diff
@@ -122,12 +122,17 @@ type PipelineResult struct {
 
 	// Value the expression used to retrieve the value
-	Value string `json:"value"`
+	Value ArrayOrString `json:"value"`
```
could you please update the commit message/PR description noting that this PR includes a breaking change to the Go libraries?

Yongxuanzhang replied:
yes sure!

ywluogg reviewed (approved):
NIce job! Everything looks good except some typos

lbernick commented on pkg/apis/pipeline/v1beta1/pipeline_types.go:
```diff
@@ -122,12 +122,17 @@ type PipelineResult struct {
+	// Type is the user-specified type of the result. The possible type
+	// is currently "string", "array" and "object" are alpha feature.
+	// +optional
+	Type ResultsType `json:"type,omitempty"`
```
since `ResultsType` already has its own docstring I'm wondering if you should remove the comment about what types are supported, since if the supported types change this comment will get out of sync? If you'd like to keep it in place, I'd suggest "The possible types are 'string', 'array', and 'object', with 'string' as the default. 'array' and 'object' types are alpha features."

lbernick commented on pkg/apis/pipeline/v1beta1/resultref.go:
```diff
@@ -52,7 +52,9 @@ const (
-var arrayIndexingRegex = regexp.MustCompile(arrayIndexing)
+
+// ArrayIndexingRegex is used to match `[int]` and `[*]`
+var ArrayIndexingRegex = regexp.MustCompile(arrayIndexing)
```
I'm wondering if we need to export this regex or if there is a parsing function it would make more sense to export

Yongxuanzhang replied:
oh that's a good point! Let me check if I can have a separate function for this

lbernick replied:
could you please avoid exporting the regexes here?

lbernick commented on pkg/reconciler/pipelinerun/resources/apply_test.go:
```diff
@@ -1079,6 +1079,66 @@ func TestApplyTaskResultsToPipelineResults(t *testing.T) {
+			Value: *v1beta1.NewArrayOrString("rae"),
+		}},
+	}, {
+		description: "multiple-array-results-multiple-successful-tasks ",
```
a bit confused, this test case is described as "multiple successful tasks" but only one pipeline task is referenced

lbernick commented:
thanks @Yongxuanzhang -- I can see you've addressed some comments, just lmk when ready for re review!

Yongxuanzhang commented:
> thanks @Yongxuanzhang -- I can see you've addressed some comments, just lmk when ready for re review!

Thank you! It is ready for another round of reviews!

lbernick commented on pkg/reconciler/pipelinerun/resources/apply.go:
```diff
@@ -203,26 +211,48 @@ func replaceParamValues(params []v1beta1.Param, stringReplacements map[string]st
 			variableParts := strings.Split(variable, ".")
 			if len(variableParts) == 4 && variableParts[0] == "tasks" && variableParts[2] == "results" {
 				taskName, resultName := variableParts[1], variableParts[3]
+				stringIdx := strings.TrimSuffix(strings.TrimPrefix(v1beta1.ArrayIndexingRegex.FindString(resultName), "["), "]")
```
I don't think this comment was addressed

Yongxuanzhang replied:
I have moved this functionality into the function `ParseResultName` in `resultre.go`, or did you mean something else?

lbernick replied:
LGTM!

Yongxuanzhang commented:
/retest

Yongxuanzhang commented:
/retest

Yongxuanzhang commented:
need to validate types between pipeline result and task result.

ywluogg commented:
/lgtm

ywluogg commented:
#4723

Yongxuanzhang commented:
/retest

lbernick commented on pkg/apis/pipeline/v1beta1/resultref.go:
```diff
@@ -56,8 +58,13 @@ const (
 var arrayIndexingRegex = regexp.MustCompile(arrayIndexing)
 
+// ArrayReplacementRegex is used to match array result reference e.g. $(tasks.taskName.results.resultName[*]))
+var ArrayReplacementRegex = regexp.MustCompile(arrayReplacement)
```
where is this variable being used?

Yongxuanzhang commented on pkg/reconciler/pipelinerun/resources/apply.go:
```diff
@@ -235,26 +235,42 @@ func replaceParamValues(params []v1beta1.Param, stringReplacements map[string]st
+							intIdx, _ := strconv.Atoi(stringIdx)
+							stringReplacements[variable] = resultValue.ArrayVal[intIdx]
+						} else {
+							variable = strings.TrimSuffix(strings.TrimSuffix(strings.TrimPrefix(variable, "$("), ")"), "[*]")
```
sure!

ywluogg commented:
/lgtm

Yongxuanzhang commented:
/retest

ywluogg commented:
/lgtm
